File: shapetrees-java-javahttp/src/main/java/com/janeirodigital/shapetrees/javahttp/JavaHttpValidatingShapeTreeInterceptor.java
package com.janeirodigital.shapetrees.javahttp;

import com.janeirodigital.shapetrees.client.http.HttpResourceAccessor;
import com.janeirodigital.shapetrees.core.*;
import com.janeirodigital.shapetrees.core.enums.HttpHeaders;
import com.janeirodigital.shapetrees.core.enums.ShapeTreeResourceType;
import com.janeirodigital.shapetrees.core.exceptions.ShapeTreeException;
import com.janeirodigital.shapetrees.core.methodhandlers.*;
import lombok.AllArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

import javax.net.ssl.SSLSession;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.net.http.HttpResponse;
import java.util.*;

/**
 * Wrapper used for client-side validation
 */
@Slf4j
public class JavaHttpValidatingShapeTreeInterceptor {

    private static final String POST = "POST";
    private static final String PUT = "PUT";
    private static final String PATCH = "PATCH";
    private static final String DELETE = "DELETE";

    @NotNull
    public java.net.http.HttpResponse validatingWrap(java.net.http.HttpRequest clientRequest, java.net.http.HttpClient httpClient, String body, String contentType) throws IOException, InterruptedException {

        ShapeTreeRequest shapeTreeRequest = new JavaHttpShapeTreeRequest(clientRequest, body, contentType);
        ResourceAccessor resourceAccessor = new HttpResourceAccessor();

        // Get the handler
        ValidatingMethodHandler handler = getHandler(shapeTreeRequest.getMethod(), resourceAccessor);
        if (handler != null) {
            try {
                Optional<DocumentResponse> shapeTreeResponse = handler.validateRequest(shapeTreeRequest);
                if (!shapeTreeResponse.isPresent()) {
                    return JavaHttpClient.check(httpClient.send(clientRequest, java.net.http.HttpResponse.BodyHandlers.ofString()));
                } else {
                    return createResponse(clientRequest, shapeTreeResponse.get());
                }
            } catch (ShapeTreeException ex){
                log.error("Error processing shape tree request: ", ex);
                return createErrorResponse(ex, clientRequest);
            } catch (Exception ex) {
                log.error("Error processing shape tree request: ", ex);
                return createErrorResponse(new ShapeTreeException(500, ex.getMessage()), clientRequest);
            }
        } else {
            log.warn("No handler for method [{}] - passing through request", shapeTreeRequest.getMethod());
            return JavaHttpClient.check(httpClient.send(clientRequest, java.net.http.HttpResponse.BodyHandlers.ofString()));
        }
    }

    private ValidatingMethodHandler getHandler(String requestMethod, ResourceAccessor resourceAccessor) {
        switch (requestMethod) {
            case POST:
                return new ValidatingPostMethodHandler(resourceAccessor);
            case PUT:
                return new ValidatingPutMethodHandler(resourceAccessor);
            case PATCH:
                return new ValidatingPatchMethodHandler(resourceAccessor);
            case DELETE:
                return new ValidatingDeleteMethodHandler(resourceAccessor);
            default:
                return null;
        }
    }

    private java.net.http.HttpResponse createErrorResponse(ShapeTreeException exception, java.net.http.HttpRequest nativeRequest) {
        return new MyHttpResponse(exception.getStatusCode(), nativeRequest, java.net.http.HttpHeaders.of(Collections.emptyMap(), (a, v) -> true), exception.getMessage());
    }

    @SneakyThrows
    private java.net.http.HttpResponse createResponse(java.net.http.HttpRequest nativeRequest, DocumentResponse response) {
        java.net.http.HttpHeaders headers = java.net.http.HttpHeaders.of(response.getResourceAttributes().toMultimap(), (a, v) -> true);
        return new MyHttpResponse(response.getStatusCode(), nativeRequest, headers, response.getBody());
    }

    private class JavaHttpShapeTreeRequest implements ShapeTreeRequest {
        private final java.net.http.HttpRequest request;
        private ShapeTreeResourceType resourceType;
        private final String body;
        private final String contentType;
        private final ResourceAttributes headers;

        public JavaHttpShapeTreeRequest(java.net.http.HttpRequest request, String body, String contentType) {
            this.request = request;
            this.body = body;
            this.contentType = contentType;
            TreeMap<String, List<String>> tm = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
            Map<String, List<String>> headerMap = this.request.headers().map();
            for (Map.Entry<String, List<String>> entry : headerMap.entrySet()) {
                tm.put(entry.getKey(), entry.getValue());
            }
            this.headers = new ResourceAttributes(tm);
        }

        @Override
        public String getMethod() {
            return this.request.method();
        }

        @Override
        public URL getUrl() {
            try {
                return this.request.uri().toURL();
            } catch (MalformedURLException ex) {
                throw new IllegalStateException("request has a malformed URL <" + request.uri() + ">: " + ex.getMessage());
            }
        }

        @Override
        public ResourceAttributes getHeaders() {
            return this.headers;
        }

        @Override
        public ResourceAttributes getLinkHeaders() {
            return ResourceAttributes.parseLinkHeaders(this.getHeaderValues(HttpHeaders.LINK.getValue()));
        }

        @Override
        public List<String> getHeaderValues(String header) {
            return this.request.headers().allValues(header);
        }

        @Override
        public String getHeaderValue(String header) {
            return this.request.headers().firstValue(header).orElse(null);
        }

        @Override
        public String getContentType() {
            return this.getHeaders().firstValue(HttpHeaders.CONTENT_TYPE.getValue()).orElse(null);
        }

        @Override
        public ShapeTreeResourceType getResourceType() {
            return this.resourceType;
        }

        @Override
        public void setResourceType(ShapeTreeResourceType resourceType) {
            this.resourceType = resourceType;
        }

        @Override
        public String getBody() {
            return this.body;
        }
    }

    @AllArgsConstructor
    private class MyHttpResponse implements java.net.http.HttpResponse {
        private int statusCode;
        private java.net.http.HttpRequest request;
        private java.net.http.HttpHeaders headers;
        private String body;

        @Override
        public int statusCode() {
            return this.statusCode;
        }

        @Override
        public java.net.http.HttpRequest request() {
            return this.request;
        }

        @Override
        public Optional<HttpResponse<String>> previousResponse() {
            return Optional.empty();
        }

        @Override
        public java.net.http.HttpHeaders headers() {
            return this.headers;
        }

        @Override
        public String body() {
            return this.body;
        }

        @Override
        public Optional<SSLSession> sslSession() {
            return Optional.empty();
        }

        @Override
        public URI uri() {
            return null;
        }

        @Override
        public java.net.http.HttpClient.Version version() {
            return null;
        }
    }
}

File: asTypescript/packages/javahttp/src/JavaHttpValidatingShapeTreeInterceptor.ts
// Corresponding shapetrees-java package: com.janeirodigital.shapetrees.javahttp
import { HttpResourceAccessor } from '@shapetrees/ttpResourceAccessor';
import * as core from 'com/janeirodigital/shapetrees';
import { HttpHeaders } from '@shapetrees/nums/HttpHeaders';
import { ShapeTreeResourceType } from '@shapetrees/nums/ShapeTreeResourceType';
import { ShapeTreeException } from '@shapetrees/xceptions/ShapeTreeException';
import * as methodhandlers from '@shapetrees/ethodhandlers';
import * as Slf4j from 'lombok/extern/slf4j';
import * as NotNull from 'org/jetbrains/annotations';
import * as SSLSession from 'javax/net/ssl';
import * as MalformedURLException from 'java/net';
import * as URI from 'java/net';
import * as URL from 'java/net';
import * as HttpResponse from 'java/net/http';
import * as util from 'java';

/**
 * Wrapper used for client-side validation
 */
@Slf4j
export class JavaHttpValidatingShapeTreeInterceptor {

   private static readonly POST: string = "POST";

   private static readonly PUT: string = "PUT";

   private static readonly PATCH: string = "PATCH";

   private static readonly DELETE: string = "DELETE";

  // @NotNull
  public validatingWrap(clientRequest: java.net.http.HttpRequest, httpClient: java.net.http.HttpClient, body: string, contentType: string): java.net.http.HttpResponse /* throws IOException, InterruptedException */ {
    let shapeTreeRequest: ShapeTreeRequest = new JavaHttpShapeTreeRequest(clientRequest, body, contentType);
    let resourceAccessor: ResourceAccessor = new HttpResourceAccessor();
    // Get the handler
    let handler: ValidatingMethodHandler = getHandler(shapeTreeRequest.getMethod(), resourceAccessor);
    if (handler != null) {
      try {
        let shapeTreeResponse: Optional<DocumentResponse> = handler.validateRequest(shapeTreeRequest);
        if (!shapeTreeResponse.isPresent()) {
          return JavaHttpClient.check(httpClient.send(clientRequest, java.net.http.HttpResponse.BodyHandlers.ofString()));
        } else {
          return createResponse(clientRequest, shapeTreeResponse.get());
        }
      } catch (ex: ShapeTreeException) {
        log.error("Error processing shape tree request: ", ex);
        return createErrorResponse(ex, clientRequest);
      } catch (ex: Exception) {
        log.error("Error processing shape tree request: ", ex);
        return createErrorResponse(new ShapeTreeException(500, ex.getMessage()), clientRequest);
      }
    } else {
      log.warn("No handler for method [{}] - passing through request", shapeTreeRequest.getMethod());
      return JavaHttpClient.check(httpClient.send(clientRequest, java.net.http.HttpResponse.BodyHandlers.ofString()));
    }
  }

  private getHandler(requestMethod: string, resourceAccessor: ResourceAccessor): ValidatingMethodHandler {
    switch(requestMethod) {
      case POST:
        return new ValidatingPostMethodHandler(resourceAccessor);
      case PUT:
        return new ValidatingPutMethodHandler(resourceAccessor);
      case PATCH:
        return new ValidatingPatchMethodHandler(resourceAccessor);
      case DELETE:
        return new ValidatingDeleteMethodHandler(resourceAccessor);
      default:
        return null;
    }
  }

  private createErrorResponse(exception: ShapeTreeException, nativeRequest: java.net.http.HttpRequest): java.net.http.HttpResponse {
    return new MyHttpResponse(exception.getStatusCode(), nativeRequest, java.net.http.HttpHeaders.of(Collections.emptyMap(), (a, v) -> true), exception.getMessage());
  }

  // @SneakyThrows
  private createResponse(nativeRequest: java.net.http.HttpRequest, response: DocumentResponse): java.net.http.HttpResponse {
    let headers: java.net.http.HttpHeaders = java.net.http.HttpHeaders.of(response.getResourceAttributes().toMultimap(), (a, v) -> true);
    return new MyHttpResponse(response.getStatusCode(), nativeRequest, headers, response.getBody());
  }

  private class JavaHttpShapeTreeRequest implements ShapeTreeRequest {

     private readonly request: java.net.http.HttpRequest;

     private resourceType: ShapeTreeResourceType;

     private readonly body: string;

     private readonly contentType: string;

     private readonly headers: ResourceAttributes;

    public constructor(request: java.net.http.HttpRequest, body: string, contentType: string) {
      this.request = request;
      this.body = body;
      this.contentType = contentType;
      let tm: TreeMap<string, List<string>> = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
      let headerMap: Map<string, List<string>> = this.request.headers().map();
      for (let entry: Map.Entry<string, List<string>> : headerMap.entrySet()) {
        tm.put(entry.getKey(), entry.getValue());
      }
      this.headers = new ResourceAttributes(tm);
    }

    override public getMethod(): string {
      return this.request.method();
    }

    override public getUrl(): URL {
      try {
        return this.request.uri().toURL();
      } catch (ex: MalformedURLException) {
        throw new IllegalStateException("request has a malformed URL <" + request.uri() + ">: " + ex.getMessage());
      }
    }

    override public getHeaders(): ResourceAttributes {
      return this.headers;
    }

    override public getLinkHeaders(): ResourceAttributes {
      return ResourceAttributes.parseLinkHeaders(this.getHeaderValues(HttpHeaders.LINK.getValue()));
    }

    override public getHeaderValues(header: string): List<string> {
      return this.request.headers().allValues(header);
    }

    override public getHeaderValue(header: string): string {
      return this.request.headers().firstValue(header).orElse(null);
    }

    override public getContentType(): string {
      return this.getHeaders().firstValue(HttpHeaders.CONTENT_TYPE.getValue()).orElse(null);
    }

    override public getResourceType(): ShapeTreeResourceType {
      return this.resourceType;
    }

    override public setResourceType(resourceType: ShapeTreeResourceType): void {
      this.resourceType = resourceType;
    }

    override public getBody(): string {
      return this.body;
    }
  }

  private class MyHttpResponse implements java.net.http.HttpResponse {

     private statusCode: number;

     private request: java.net.http.HttpRequest;

     private headers: java.net.http.HttpHeaders;

     private body: string;

    override public statusCode(): number {
      return this.statusCode;
    }

    override public request(): java.net.http.HttpRequest {
      return this.request;
    }

    override public previousResponse(): Optional<HttpResponse<string>> {
      return Optional.empty();
    }

    override public headers(): java.net.http.HttpHeaders {
      return this.headers;
    }

    override public body(): string {
      return this.body;
    }

    override public sslSession(): Optional<SSLSession> {
      return Optional.empty();
    }

    override public uri(): URI {
      return null;
    }

    override public version(): java.net.http.HttpClient.Version {
      return null;
    }

    public constructor(statusCode: number, request: java.net.http.HttpRequest, headers: java.net.http.HttpHeaders, body: string) {
      this.statusCode = statusCode;
      this.request = request;
      this.headers = headers;
      this.body = body;
    }
  }
}
